### PR TITLE
WIP: Generated using the latest features from MoveIt

### LINKED ISSUE #72: Noetic Release (0.8.0)
# Noetic Release (0.8.0)

`franka_description` was released: https://index.ros.org/p/franka_description/#noetic

Before we cut the release we should regenerate this based on the master version of the moveit setup assistant.  @rhaschke 

### Checklist
- [ ] Update based on master branch of moveit setup assistant
- [ ] Make `noetic-devel` branch
- [ ] Run ROS buildfarm prerelease test
- [ ] Update changelogs
- [ ] Bloom
- [ ] Write discourse post

### PR DESCRIPTION
**Pull request by tylerjw**
# Description

This was created by generating using the master branch of MoveIt.  There are a handful of files that were not generated (i.e. `config/panda_arm_hand.srdf.xacro`), I'd like to document how they are created and what their use is in the README.  

Lastly, I chose two new poses that are as close to the original ones as I could make and stay out of collision.  I didn't include the `transport` pose because I was not able to make that pose and stay out of collision.  I've included images of each pose here:
panda_extended
panda_ready
panda_transport

Lastly, this should be tested with moveit_tutorials to make sure it doesn't break any of those.

## Notes to reviewers

Here are the things that need to be done before this is ready:
* [ ] Identify and document changes made after generating from MSA
* [ ] Discuss new poses
* [ ] Test with moveit_tutorials

### CODE REVIEW

**Comment by v4hn** [comment]
The `transport` pose is the configuration used to ship the robot in its original box.
That's the main reason why it was part of the configuration.

Yes, it's in collision with the current model and this was discussed somewhere before. The MoveIt-compatible trajectory controller can't move the robot to this pose because the internal self-collision check will fail as well. Apparently they have a separate internal routine to move to this pose.

**Comment by tylerjw** [comment]
I checked through the commits since the last release and this only backs out one change.  There is a PR to add the transport pose to one of the files.  This specific file I think we can safely remove, this I explain below.

This brings us to a point of discussion.  There are two sorts of files in this repo that are not generated by moveit_setup_assistant.  The first is xacro files that were copied versions of the old robot description.  I believe these can be safely removed.  

The next are files that support specific tutorials and were manually added here.  An example of that is the `lerp_planning_pipeline.launch.xml` and related changes.  I'm not sure what the right approach is for these sorts of files.  I think it makes sense to document what these files are so that in the future if someone needs to generate this package to update it they know what to do about those files.

That or we should move these launch and config files out of this repo and into the tutorials repo to go alongside their respective tutorials.  A third option would be to extend the setup assistant to generate these files if they are actually general-purpose launch files that would be useful for other robot setups.

@rhaschke @v4hn do you have thoughts on how we should best maintain these manually added files?  I am working through identifying them and trying to validate the functionality of this PR with tutorials.

**Comment by tylerjw** [comment]
Here is a list of the tutorials on master of moveit_tutorials with my notes.  I'll add notes as I work through each one comparing it to the current version of panda_moveit_config.

* [x] Getting Started
* [x] MoveIt Quickstart in RViz
* [x] Move Group C++ Interface
  - The ready state is slightly different causing some of the paths to be changed slightly, but overall this tutorial works.  After this is submitted the images and video should be updated.
  - Noticed the generated rviz file doesn't have the `rviz_visual_tools` gui panel.  I'm going to put that file back and note it as not a generated file.
* [x] Move Group Python Interface
  - [Python3 migration](https://github.com/ros-planning/moveit_tutorials/pull/536) is needed to run this on Noetic.
  - Tutorial works the same with the new and old configuration.
  - The position of the box using both the new and old configuration is odd.
  - offset_box_old
* [x] MoveIt Commander Scripting
* [x] Robot Model and Robot State
* [x] Planning Scene
* [x] Planning Scene Monitor
* [x] Planning Scene ROS API
  - Object spawns inside the gripper.  See image.
  - The last step "Remove the object from the collision world" doesn't seem to work as the object is not removed.  This seemingly broken behavior exists with the old config too.
  - planning_scene_api_tutorial_collision
* [x] Motion Planning API
* [x] Motion Planning Pipeline
* [ ] Creating MoveIt Plugins
* [ ] Visualizing Collisions
  - This one is currently broken on the current setup.
* [x] Time Parameterization
* [ ] Planning with approximated Constraint Manifolds
* [x] Pick and Place
  - This works as-is but generates a much crazier path because of the new collision model for the robot.
* [ ] MoveIt Grasps
* [ ] MoveIt Task Constructor
* [ ] MoveIt Deep Grasps
* [ ] Subframes
  - This is broken with the old configuration with master of MoveIt on Noetic.  The object spawns floating in space between the grippers.  The grippers should probably be closed to "hold" the cylinder but that is probably not the primary cause of the issues with this tutorial.  Each of the steps fails to plan, I'm not sure why.  @felixvd have you seen this?
  - subframes_tutorial
* [x] MoveItCpp Tutorial
* [x] MoveIt Setup Assistant
* [x] URDF and SRDF
* [x] Low Level Controllers
* [ ] Preception Pipeline Tutorial
* [ ] IKFast Kinematics Solver
* [ ] TRACK-IK Kinematics Solver
* [ ] Kinematics Configuration
* [x] Custom Constraint Samplers
* [ ] OMPL Planner
* [ ] CHOMP Planner
* [ ] STOMP Planner
* [ ] TrajOpt Planner
* [ ] Planning Adapter Tutorials
* [ ] Joystick Control Teleoperation
* [ ] Realtime Arm Servoing
* [ ] Benchmarking
  - Needs https://index.ros.org/p/warehouse_ros_mongo/github-ros-planning-warehouse_ros_mongo which isn't released for Noetic yet.  The instructions need to be updated to include building this from source until this is released.
  - I'm confused on how to run the first section called [Example](https://ros-planning.github.io/moveit_tutorials/doc/benchmarking/benchmarking_tutorial.html?highlight=collision_scene_example#example).  It feels like it is missing instructions for loading the `Kitechen1` scene objects.  Maybe I did something wrong?
  - [x] https://github.com/ros-planning/warehouse_ros_mongo/pull/35 is needed for running on Noetic.
* [ ] Integration/Unit Tests

## Generated then modified files
* `launch/moveit.rviz` - restored to state on melodic-devel

## Custom Files added for Tutorials
* placeholder

## Modifications needed for tutorials
* [ ] Update video and images for "Move Group C++ Interface"
* [x] Fix the position of the box in the "Move Group Python Interface" so it is in between the fingers.
* [ ] Update MSA tutorial to include instructions for the default positions of the end effector.
* [ ] Clarify the "Example" section of "Benchmarking" to include instructions for the collision scene objects.
* [ ] Add instructions to "Benchmarking" for building warehouse_ros_mongo from source until it is released to Noetic.
* [x] Add note to tutorials README.md that rosdoc_lite is needed to run the build_locally.sh script.
* [x] Fix "Planning Scene ROS API" tutorial to spawn the object in the gripper and remove it from the world in the last step.
* [x] Fix "Visualizing Collisions"
* [x] Fix "Subframes" tutorial

**Comment by v4hn** [comment]
> The first is xacro files that were copied versions of the old robot description.
> I believe these can be safely removed.

This is not the "old robot description", but the srdf file that is split in multiple sub-files. If you drop the ability to load the arm without gripper from the updated package, they should be removed. Some users might be interested in using the arm with other endeffectors, but they should usually generate their own moveit config anyway. So unless the distinction is required for some tutorial, I assent to revert to a single srdf file.

> A third option would be to extend the setup assistant to generate these files if they are actually general-purpose launch files that would be useful for other robot setups.

If some files are general-purpose, they should be added to the templates.
The LERP planner is created in the tutorial though, so I would just keep it in the panda config with an appropriate comment at the top of the file.

> That or we should move these launch and config files out of this repo and into the tutorials repo to go alongside their respective tutorials. 

The `planning_pipeline` "plugin-like" structure does not allow for this and we discussed some time ago in some issue whether we want to extend it to do so. At least my personal opinion was that it does not make sense to add plugin mechanisms to launch files and people can directly include them in their own launch files.
However, I'm not sure how much overhead that would create for the tutorial's launch file, which should definitely be as simple as possible to support new users.

**Comment by tylerjw** [comment]
> This is not the "old robot description", but the srdf file that is split in multiple sub-files. If you drop the ability to load the arm without gripper from the updated package, they should be removed. Some users might be interested in using the arm with other endeffectors, but they should usually generate their own moveit config anyway. So unless the distinction is required for some tutorial, I assent to revert to a single srdf file.

I believe this functionality is now provided by the description files in the upstream franka package.  From what I could tell our versions of these do not differ in functionality from the official ones now.  I have yet to find a tutorial that used them, or tested loading without the gripper using the upstream description files.

**Comment by tylerjw** [comment]
Today I went through and investigates some of the easier to test tutorials.  I'll need to work through the rest of them but unfortunatly I found some that don't work with the old config and the master branch of MoveIt on Noetic.  Some of these might be easy fixes but I'd don't believe they should block this change and should hopefully be unaffected by this change.

**Comment by v4hn** [comment]
> > This is not the "old robot description", but the srdf file that is split in multiple sub-files.
> 
> I believe this functionality is now provided by the description files in the upstream franka package. From what I could tell our versions of these do not differ in functionality from the official ones now. I have yet to find a tutorial that used them, or tested loading without the gripper using the upstream description files.

Are we talking about the same files? [hand.xacro in this repository](https://github.com/ros-planning/panda_moveit_config/blob/melodic-devel/config/hand.xacro#L9) provides **SRDF** specifications, [hand.xacro in franka_ros](https://github.com/frankaemika/franka_ros/blob/kinetic-devel/franka_description/robots/hand.xacro#L12) provides **URDF** specs. These files do not have anything in common aside from their name.

**Comment by tylerjw** [comment]
> Are we talking about the same files? [hand.xacro in this repository](https://github.com/ros-planning/panda_moveit_config/blob/melodic-devel/config/hand.xacro#L9) provides **SRDF** specifications, [hand.xacro in franka_ros](https://github.com/frankaemika/franka_ros/blob/kinetic-devel/franka_description/robots/hand.xacro#L12) provides **URDF** specs. These files do not have anything in common aside from their name.

I didn't realize this.  I figured they provided the same utility.  My reasoning was based on searching for files that used these various files in the panda_moveit_config repo and in all cases it seemed like they were referring to the one in franka_description.  I'll put those back.

File: config/panda_arm.xacro [hunk 1]
```diff
@@ -11,31 +11,22 @@
...
-      <joint name="panda_joint5" value="0" />
-      <joint name="panda_joint6" value="1.571" />
-      <joint name="panda_joint7" value="0.785" />
+        <joint name="panda_joint1" value="0"/>
+        <joint name="panda_joint2" value="-0.785"/>
+        <joint name="panda_joint3" value="0"/>
+        <joint name="panda_joint4" value="-2.356"/>
+        <joint name="panda_joint5" value="0"/>
+        <joint name="panda_joint6" value="1.571"/>
+        <joint name="panda_joint7" value="-0.7855"/>
```
Comment (henningkayser) [inline config/panda_arm.xacro]: Why are the default poses changing? They shouldn't be affected by running MSA.

File: launch/demo.launch [hunk 2]
```diff
@@ -1,39 +1,40 @@
...
-  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
-    <param name="/use_gui" value="$(arg use_gui)"/>
-    <rosparam param="/source_list">[/move_group/fake_controller_joint_states]</rosparam>
+  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" unless="$(arg use_gui)">
+    <rosparam param="source_list">[move_group/fake_controller_joint_states]</rosparam>
+  </node>
+  <node name="joint_state_publisher" pkg="joint_state_publisher_gui" type="joint_state_publisher_gui" if="$(arg use_gui)">
+    <rosparam param="source_list">[move_group/fake_controller_joint_states]</rosparam>
   </node>
-  <node name="joint_state_desired_publisher" pkg="topic_tools" type="relay" args="joint_states joint_states_desired" />
```
Comment (henningkayser) [inline launch/demo.launch:36]: Not sure if we should really remove this

**Review by henningkayser — COMMENTED** [review]
Overall, I really think it makes a lot of sense to keep configs up-to-date with the MSA, imo it's just strainge if this is not possible. I'm just not sure what setups we would break by doing this. Of course we should keep modifications like with the srdf files and default poses consistent.

File: config/panda_arm.xacro (hunk 1 above)
Comment (tylerjw) [reply config/panda_arm.xacro]: for the upstream collision scene

**Comment by JafarAbdi** [comment]
Note that I disabled the collision between panda_link8 & panda_link6 + panda_link7 otherwise hand movement is very restricted causing some of the tutorials to not work

**Comment by davetcoleman** [comment]
This seems overdue for a merge, @tylerjw can we push this in?

**Comment by tylerjw** [comment]
@davetcoleman I abandoned this because I didn't have the time to finish it.  There are a handful of tutorials it needs to be tested with and either this or the tutorials need to be fixed.  I've left documentation of all of that above.

**Comment by davetcoleman** [comment]
@tylerjw by "abandoned" you mean set it aside, correct? You would support someone else taking this on, as its still important right?

**Comment by tylerjw** [comment]
I mean that I stopped working on it.  I have no issue with someone else taking it up.  I think that person will need to have a good understanding of the uses of this package (probably more than me) and all the tutorials or they risk getting stuck just as I did.

**Comment by rickstaa** [comment]
@tylerjw I made some changes that make sure that the `panda_control_moveit_rviz.launch` file can be used again to control the real robot using moveit. I created a pull request on your branch https://github.com/tylerjw/panda_moveit_config/pull/1.

**Comment by rickstaa** [comment]
@tylerjw First of all thanks a lot for creating this pull request. @davetcoleman I'm happy to help. I made some changes to ensure that the `panda_control_moveit_rviz.launch` file can be used again to control the real robot using move it. In addition, I created a pull request on your branch https://github.com/tylerjw/panda_moveit_config/pull/1. I will create additional pull requests, if I run into more errors while using the [tylerjw:p/update_latest](https://github.com/tylerjw/panda_moveit_config/tree/p/update_latest) branch.

**Comment by rickstaa** [comment]
@tylerjw I updated the robot_description xacro url since the upstream was changed (see https://github.com/tylerjw/panda_moveit_config/pull/3).

**Comment by rickstaa** [comment]
@tylerjw Thanks for merging the last pull request I created a new pull request with two fixes that are necessary (see https://github.com/tylerjw/panda_moveit_config/pull/6). Further, there is a dependency pull request in https://github.com/tylerjw/panda_moveit_config/pull/4. I will now try to debug why the controllers are not working when launching the `franka_example_controllers.move_start.launch` script.

**Comment by rickstaa** [comment]
I created https://github.com/frankaemika/franka_ros/issues/162 to make the Franka Emika team aware of the restrictive hand movement problem. @JafarAbdi please free to add any points I missed or misunderstood in your original explanation to the issue.

**Comment by rickstaa** [comment]
Last week I gave this feature another try while keeping the points that @tylerjw mentioned in https://github.com/ros-planning/panda_moveit_config/pull/74#issuecomment-703129249 in mind. 

To do this, I first created an orphan branch and created the `panda_moveit_config` again using the [latest moveit_setup_assistant](https://github.com/ros-planning/moveit/tree/master/moveit_setup_assistant). I did this both to get all the latest changes from the `moveit_setup_assistant` into this repository and to get a sense for which features were missing in the MSA. I then manually compared the new files with those in the [melodic-devel branch](https://github.com/ros-planning/panda_moveit_config/tree/melodic-devel) keeping notes and create pull request where needed. The result of this can be found [in this pull request](https://github.com/rickstaa/panda_moveit_config/pull/26).

https://github.com/rickstaa/panda_moveit_config/pull/26 is nearly ready to be merged onto #74, but I need some help applying some last pull requests (see the sections below).

## Pull requests that first need to be reviewed

Below are some pull request that I did not yet merge since they first need to be reviewed by the @ros-planning team.
- [ ] https://github.com/rickstaa/panda_moveit_config/pull/54 I did not yet merge this pull request since I'm not sure if this change is still required with [franka_ros](https://github.com/frankaemika/franka_ros) v0.8.0. Please let me know if somebody knows the answer. I will try to perform some tests on the real robot later to find out.

## Upstream pull requests that need to be merged

The pull requests below were created  on the [frankaemika/franka_ros](https://github.com/frankaemika/franka_ros) repository and need to be merged before the Noetic version of the [ros-planning/panda_moveit_config](https://github.com/ros-planning/panda_moveit_config) can be released.

- [ ] https://github.com/frankaemika/franka_ros/pull/151 needs to be merged to get rid of the ` Link panda_rightfinger has visual geometry but no collision geometry. Collision geometry will be left empty. Fix your URDF file by explicitly specifying collision geometry.` warning.
- [ ] https://github.com/frankaemika/franka_ros/pull/181 needs to be merged. Such that we can also use the `joint_position/JointTrajectoryController` controller in the gazebo simulation. We can also of course switch to a `joint_effort/JointTrajectoryController` but then we need to tune the PID gains (see comment above). 
- [ ] https://github.com/frankaemika/franka_ros/pull/173 needs to be merged before MoveIt can correctly control the gripper.
- [ ] If https://github.com/frankaemika/franka_ros/pull/154 is merged, we can use `use_mesh` to improve the self-collision behaviour. aa24fae70f4a62ec63fc72ecbb0aacdb4283738e is then not needed anymore.

**Comment by rhaschke** [comment]
Thanks a lot, @rickstaa, for pushing this forward. A new release of panda_moveit_config is long overdue!
I had a look at your branch https://github.com/rickstaa/panda_moveit_config/pull/26 and cleaned it a little bit, following these rules (see [here](https://github.com/rhaschke/panda_moveit_config/tree/noetic-devel-rickstaa) for the result):
- commit messages shouldn't include PR IDs from your fork
- fixups due to changes in MSA behavior were squashed directly into the base commit

Regarding your questions:
- I suggest including examples for other planning pipelines even though these pipelines are currently not operational or released. Otherwise, it will become hard to re-add those files later from scratch.
- For the Gazebo simulation, I suggest to use `franka_gazebo` if that's closer to the real robot. Does that require changes in the tutorials as well?
- If the existing files differ to the MSA-generated ones: Keep the generated ones, if generic files (e.g. planning pipeline settings) are concerned. But keep robot-specific settings like joint limits etc. For those, MSA can only use the defaults from the URDF.
- @Le-Li-17 please have a look at the long list of PRs filed to Franka. We strongly believe that the URDF model released with version 0.8.1 still has severe issues strongly limiting the motion range of the simulated robot. See https://github.com/ros-planning/panda_moveit_config/pull/35#pullrequestreview-390715967

**Comment by rickstaa** [comment]
@rhaschke Thanks a lot for helping me implement this new release.

Thanks for rebasing my commits they look a lot cleaner! I will use https://github.com/rhaschke/panda_moveit_config/tree/noetic-devel-rickstaa and adhere to these rules moving forward. Are these rules written down somewhere in a `contributing.md`. Or are they requested when people make a pull request?


> I suggest including examples for other planning pipelines even though these pipelines are currently not operational or released. Otherwise, it will become hard to re-add those files later from scratch.

Are you okay with adding them commented out since otherwise, they throw errors (see https://github.com/rickstaa/panda_moveit_config/blob/f4e1c046d7ca88d9e06f0969ed81cfda62fb2d0b/launch/move_group.launch#L64-L70).

> For the Gazebo simulation, I suggest to use franka_gazebo if that's closer to the real robot. Does that require changes in the tutorials as well?

Good point I did not yet look into this. I check and track all the changes that need to be performed to the `moveit_tutorials` in https://github.com/rickstaa/panda_moveit_config/pull/26.

> @Le-Li-17 please have a look at the long list of PRs filed to Franka. We strongly believe that the URDF model released with version 0.8.1 still has severe issues strongly limiting the motion range of the simulated robot. See Update SRDF for collision model #35 (review)

I'm trying to fix this with @frankaemika. https://github.com/frankaemika/franka_ros/issues/162 is related to this pull request might people want to chip in.

**Comment by rhaschke** [comment]
> I will adhere to these rules moving forward. Are these rules written down somewhere in a `contributing.md`?

I just formulated these rules in a common-sense fashion to answer you questions in a generic way. I hope other maintainers agree.

> > I suggest including examples for other planning pipelines even though these pipelines are currently not operational.
> 
> Are you okay with adding them commented out since otherwise, they throw errors.

I wouldn't include them by default into `move_group.launch`, but consider them as optional custom pipelines to be included manually via the `pipeline` argument.

**Comment by rickstaa** [comment]
@rhaschke I think we can merge the changes in https://github.com/rickstaa/panda_moveit_config/pull/29 (STOMP algorithm) into https://github.com/rhaschke/panda_moveit_config/tree/noetic-devel-rickstaa even though there appears [to be an unsolved upstream issue](https://github.com/rickstaa/panda_moveit_config/pull/29#issuecomment-906377648)?

**Comment by rhaschke** [comment]
Sure! Even though the planner is failing, the config is probably still fine.

**Comment by rickstaa** [comment]
@rhaschke Just to double-check. The CHOMP and LERP planners are not supposed to work with the panda hand Parallel Jaw gripper?

**Comment by rickstaa** [comment]
@rhaschke I think I just created the last pull request needed to release the Noetic branch (i.e. https://github.com/rhaschke/panda_moveit_config/pull/8). Just to be sure, tomorrow I will perform some tests on the real robot to see if https://github.com/rickstaa/panda_moveit_config/pull/54 is still needed in ROS Noetic. After that, the Noetic branch is ready to be released when the following upstream PR are merged:

- [ ] frankaemika/franka_ros#151 needs to be merged to get rid of the `Link panda_rightfinger has visual geometry but no collision geometry. Collision geometry will be left empty. Fix your URDF file by explicitly specifying collision geometry.` warning. 
- [ ] frankaemika/franka_ros#181 needs to be merged for us to use the `position_joint_trajectory_controller`.
- [ ] frankaemika/franka_ros#173 needs to be merged before MoveIt can correctly control the gripper. Currently, MoveIt can only open and close the gripper but not set a desired width.
- [ ]  frankaemika/franka_ros#154 could improve the self-collision behaviour. When this pull request is merged, we can 'use_collision_mesh' arg to let MoveIt plan with the more accurate collision meshes. In that case aa24fae is then not needed anymore. If we however decide to do this we have to keep in mind that MoveIt can plan paths that are not possible on the real robot.
- [ ] The Noetic branch has been created using the `moveit_setup_assistant` found in [tree/82378877879f7d40727819b1c3345eeb4308d98b](https://github.com/ros-planning/moveit/tree/82378877879f7d40727819b1c3345eeb4308d98b). The current MoveIt debian package does not include the `pilz_industrial_motion_planner`. As a result, MoveIt first has to create a new release before it can be used with the debian packages.

## Migration guide

For future reference, the following steps were performed to release the Noetic branch:

1. I created the initial panda robot configuration using the [moveit_setup_assistant](https://github.com/ros-planning/moveit/tree/master/moveit_setup_assistant) (see 8566b70f5e464e6fcb1c7d5664ff6e6e5202876e). In this branch, [tree/82378877879f7d40727819b1c3345eeb4308d98b](https://github.com/ros-planning/moveit/tree/82378877879f7d40727819b1c3345eeb4308d98b) was used. Similar to https://github.com/ros-planning/panda_moveit_config/pull/74 I changed some posed to prevent self collisions.

2. I modified the `panda.srdf` file that is created by the `moveit_setup_assistant` such that we can programmatically enable the gripper (see f0c2bca91d14a5b4cd7192c189d253f4d1fb6407). This needs to be done, since the `moveit_setup_assistant` currently doesn't propagate `xacro` arguments you supply to the `urdf.xacro` file (see [this issue](https://github.com/ros-planning/moveit/issues/1691)). We therefore need to manually add a way to enable or disable the gripper. This can be done by creating a `panda.srdf.xacro` file.

3. I replaced the `sensor_3d.yaml` with the example files in the [perception_pipeline](https://ros-planning.github.io/moveit_tutorials/doc/perception_pipeline/perception_pipeline_tutorial.html?highlight=perception%20tutorial) tutorial. Also edit the `sensor_manager.launch.xml` file to point to the right configuration files (see 2aaa01d65491d02ef6c095f7bc7b13a8f52d1be6).

4. I added the `rviz_tutorial` flag. While doing this, I made sure that I copied the `moveit.rviz` and `empty.rviz` files from the previous ROS version. Additionally, I added the `moveit_visual_tools` package as a dependency (see 1d55c9851bf388e3e2c29d7ed7390c8edce32f22).

5. Similar to df0cbd8  I had to disable the collision checking between joint{5, 6} and joint8 since the new collision geometries that were added to https://github.com/frankaemika/franka_ros/releases/tag/0.8.0 are to coarse (see aa24fae70f4a62ec63fc72ecbb0aacdb4283738e). This should be fixed upstream by, for example, by improving the collision geometries or by giving users the ability to use meshes instead (https://github.com/frankaemika/franka_ros/pull/154).

6. I added the [panda_control_moveit_rviz.launch](https://github.com/rickstaa/panda_moveit_config/blob/noetic-devel-update/launch/panda_control_moveit_rviz.launch) which can be used to control the real robot (see 36d7c9f3b20f5a906b256313f3bbc50145ded820).

7. I added the [TrajOpt](https://ros-planning.github.io/moveit_tutorials/doc/trajopt_planner/trajopt_planner_tutorial.html) files that were found in the [melodic-devel](https://github.com/ros-planning/moveit_tutorials/tree/melodic-devel) branch to the repository ( see bc2bbc75620aaf01eef092fcd7301c22bef4f253 and 88fb27864bfea0f3056a1d1a39771495583320b6). 

8. I added the [STOMP](https://ros-planning.github.io/moveit_tutorials/doc/stomp_planner/stomp_planner_tutorial.html) files that were found in the [melodic-devel](https://github.com/ros-planning/moveit_tutorials/tree/melodic-devel) branch to the repository ( see 133d524ecd893180024c1d193b49b010f52c96ef) . 

9. I updated the `join_limits.yaml` file that was created by the MSA to enforce jerk limits. This was done by adding acceleration limits, since MoveIt does not yet support jerk limits (see 89d9b44cdddaca4746111795c8483e2370da5d34).

10. I added the files of the [LERP](https://ros-planning.github.io/moveit_tutorials/doc/creating_moveit_plugins/plugin_tutorial.html?highlight=lerp) planner that is created in the [creating-moveit-plugins](https://ros-planning.github.io/moveit_tutorials/doc/creating_moveit_plugins/plugin_tutorial.html?highlight=lerp#creating-moveit-plugins) tutorial (see 85756cd3d16c6d520274d40018bacc4c1779b33b). These files were copied from the [melodic-devel](https://github.com/ros-planning/moveit_tutorials/tree/melodic-devel) branch.

11. I added the `demo_chomp.launch` and `ompl-chomp_planning_pipeline.launch.xml` files needed by the [chomp_planner_tutorial](https://ros-planning.github.io/moveit_tutorials/doc/chomp_planner/chomp_planner_tutorial.html) tutorial. I also updated the chomp config file to the one found in the [melodic-devel](https://github.com/ros-planning/moveit_tutorials/tree/melodic-devel) branch (see 4c390ac0b2370aa5015e1f84c74a2818deaf9330 and c2ccca2d3f043520cfd948b7033737329b7281d5).

12. I added the `demo_gazebo.launch` launch file which spawns a simulated Panda robot in Gazebo (see c46c81c8a1e0da1288d8ef4462b8daa10df37213). This launch file uses the new [franka_gazebo](https://github.com/frankaemika/franka_ros/tree/develop/franka_gazebo) package that was introduced in [franka_ros v0.8.1](https://github.com/frankaemika/franka_ros/releases/tag/0.8.1).


### Files that were manually created

#### Config folder

-   panda.srdf.xacro
-   hand.xacro
-   panda_arm.xacro
-   lerp_planning.yaml
-   stomp_planning.yaml
-   trajopt_planning.yaml
-   panda_controllers.yaml
-   panda_gripper_controllers.yaml

#### launch

-   demo_stomp.launch (Was not present in the melodic-devel branch).
-   demo_chomp.launch (Not used in the moveit_tutorials).
-   demo_lerp.launch (Used in the moveit_tutorials).
-   lerp_planning_pipeline.launch.xml
-   ompl-chomp_planning_pipeline.launch.xml
-   panda_gripper_moveit_controller_manager.launch.xml
-   run_benchmark_trajopt.launch
-   stomp_planning_pipeline.launch.xml
-   trajopt_planning_pipeline.launch.xml

## Moveit_tutorial changes

I found the following changes that need to be performed to the [moveit_tutorials](https://github.com/ros-planning/moveit_tutorials):

- [ ]  [setup_assistant_tutorial.html](https://ros-planning.github.io/moveit_tutorials/doc/setup_assistant/setup_assistant_tutorial.html):
    -  [ ] Should have updated pictures (i.e., menu order was changed).
    -  [ ] Panda description file was changed to be `/franka_description/robots/panda.urdf.xacro` the hand is now enabled by using the `hand:=true` xacro parameter.
    -  [ ] I think the documentation should mention the `panda_finger_joint2` as the passive joint.
    -  [ ] The virtual joint doesn't appear to be needed in the panda_arm group. Maybe remove it as it will cause errors when using a gazebo simulation?
-  [ ] Maybe the [panda_control_moveit_rviz.launch] file should be documented in the documentation.
-  [ ] The `demo_chomp.launch` file was added to the [panda_moveit_config](https://github.com/ros-planning/panda_moveit_config/) repository, but this file is not used in the moveit_tutorials. 
-  [ ] The lerp planner documentation talks about a `lerp_example.launch` file but this file is not included I think this should be changed to refer the `demo_lerp.launch` file.

## Moveit_setup_assistant changes

I found several improvements that can be made to the MSA.

-   [ ] Maybe we can add the ability to use action services as controllers (i.e. https://github.com/rickstaa/panda_moveit_config/blob/47073b308a66a322146c7544f943533575f99d9e/config/panda_gripper_moveit_controllers.yaml#L14-L20).
-   [ ] Maybe we should move `controllers_list` from `ros_controllers.yaml` into its own file as these are only needed for MoveIt while the settings in the `ros_controllers.yaml` file should only contain settings related to the ROS control (see https://github.com/rickstaa/panda_moveit_config/pull/45 for a example).
-   [ ] The code that loads the `robot_description` in the [gazebo.launch` template](https://github.com/ros-planning/moveit/blob/a0ee2020c4a40d03a48044d71753ed23853a665d/moveit_setup_assistant/templates/moveit_config_pkg_template/launch/gazebo.launch#L15) from:
    ```yaml 
    <param name="robot_description" textfile="$(arg urdf_path)" />
    ````
    to
    ```yaml
    <param name="robot_description" command="$(find xacro)/xacro $(arg urdf_path) hand:=$(arg load_gripper) gazebo:=true"/>
    ```
    if a `xacro` file is supplied by the user. 
- [ ] As the https://github.com/frankaemika/franka_ros was changed, the code that creates the gazebo simulation should be adjusted. The https://github.com/frankaemika/franka_ros repo now includes all the components to create a gazebo simulation, but loads the robot onto the `/panda` namespace. As a result, the current templates don't work anymore. We should apply the changes found in https://github.com/rickstaa/panda_moveit_config/pull/40 to the templates in the MSA.  
- [ ] The STOMP templates should be added to the MSA.
- [ ] The `stomp_demo` and `chomp_demo` files are not created by the MSA. Maybe we should add these files?
- [ ] The `ompl-chomp_planning_pipeline.launch.xml` should be added to the MSA (see https://ros-planning.github.io/moveit_tutorials/doc/chomp_planner/chomp_planner_tutorial.html#using-chomp-as-a-post-processor-for-ompl).
- [ ] The CHOMP algorithm contains several commented out parameters that might be removed (see https://github.com/ros-planning/moveit/blob/d57721cf8503fe893347196cc1e1f9249ec57777/moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_parameters.h#L81-L88). 
- [ ] Maybe we also want to add the remaping found in https://github.com/rickstaa/panda_moveit_config/pull/54 into the `moveit_setup_assistant` template.
- [ ] The CHOMP `ridge_factor` parameter should be changed to `0.0` since this is the default.
- [ ] The CHOMP `enable_failure_recovery` parameter should be changed to `false` since this is the default.

## Other upstream problems that still exist

- [ ] I did not test the workings of the TrajOpt planner since it did not build on my system (see https://github.com/ros-planning/moveit/pull/1626#issuecomment-933393754).
- [ ] The STOMP planner seems to fail on ROS Noetic due to a [upstream bug](https://github.com/ros-industrial/industrial_moveit/issues/53).

## Other things we need to do

- [ ] We need to re-test all the moveit_tutorials with the new Noetic branch.

**Comment by rhaschke** [comment]
Big thanks for your work, @rickstaa :tada:
I have pushed the branch to https://github.com/ros-planning/panda_moveit_config/tree/noetic-devel

Could you please open a new issue (e.g. titled `Noetic release`) with all the info in https://github.com/ros-planning/panda_moveit_config/pull/74#issuecomment-947646655?
Closing here.

**Comment by simonschmeisser** [comment]
Note that STOMP moved to a separate repo some while ago and the issue was just fixed: https://github.com/ros-industrial/stomp_ros/pull/26

**Comment by rickstaa** [comment]
@rhaschke I created https://github.com/ros-planning/panda_moveit_config/issues/89. Feel free to request changes to the TODOS that are in there.

**Comment by rickstaa** [comment]
@simonschmeisser Thanks a lot for letting me know I will re-test the STOMP planner.

**Comment by rickstaa** [comment]
@simonschmeisser Your right the upstream issue has been solved thanks for letting me know.

**Pull request closed without merging**